File: frontend/src/types/review.ts
import * as z from 'zod'
import { UserSummarySchema } from '@/types/common'
import { RuleMatchSchema } from '@/types/matching'
import { TransactionSchema } from '@/types/transactions'

const ReviewIssueSchema = z.enum(['all', 'category', 'contract'])
export type ReviewIssue = z.infer<typeof ReviewIssueSchema>

export const AssignmentReviewFiltersSchema = z
  .object({
    issue: ReviewIssueSchema.default('all'),
    owner_id: z.number().int().positive().optional(),
    account_id: z.number().int().positive().optional(),
    contract_id: z.number().int().positive().optional(),
    q: z.string().optional(),
    page: z.number().int().positive().default(1),
    page_size: z.number().int().positive().max(100).default(50)
  })
  .strict()

export type AssignmentReviewFilters = z.infer<typeof AssignmentReviewFiltersSchema>

const AssignmentReviewRowSchema = z
  .object({
    transaction: TransactionSchema,
    account_name: z.string(),
    owner: UserSummarySchema,
    category_match: RuleMatchSchema,
    contract_match: RuleMatchSchema,
    issues: z.array(z.enum(['category', 'contract']))
  })
  .strict()

export const AssignmentReviewPageSchema = z
  .object({
    items: z.array(AssignmentReviewRowSchema),
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total: z.number().int().nonnegative(),
    total_pages: z.number().int().positive()
  })
  .strict()

export type AssignmentReviewPage = z.infer<typeof AssignmentReviewPageSchema>

export const AssignmentBulkUpdateSchema = z
  .object({
    transaction_ids: z.array(z.number().int().positive()).min(1).max(500),
    set_category: z.boolean(),
    category_id: z.number().int().positive().nullable().optional(),
    set_contract: z.boolean(),
    contract_id: z.number().int().positive().nullable().optional()
  })
  .strict()

export type AssignmentBulkUpdate = z.infer<typeof AssignmentBulkUpdateSchema>

export const AssignmentBulkUpdateResponseSchema = z
  .object({ updated: z.number().int().nonnegative() })
  .strict()

export const PatternPreviewRequestSchema = z
  .object({
    patterns: z.string(),
    owner_id: z.number().int().positive().optional(),
    start_date: z.iso.date().nullable().optional(),
    end_date: z.iso.date().nullable().optional()
  })
  .strict()

export type PatternPreviewRequest = z.infer<typeof PatternPreviewRequestSchema>

export const PatternPreviewResponseSchema = z
  .object({
    total: z.number().int().nonnegative(),
    examples: z.array(
      z
        .object({
          id: z.number().int().positive(),
          recipient: z.string(),
          subject: z.string(),
          date_issue: z.iso.date(),
          account_name: z.string()
        })
        .strict()
    )
  })
  .strict()

export type PatternPreview = z.infer<typeof PatternPreviewResponseSchema>

const TransferReviewStatusSchema = z.enum(['suggested', 'confirmed', 'rejected'])
export type TransferReviewStatus = z.infer<typeof TransferReviewStatusSchema>

export const TransferReviewFiltersSchema = z
  .object({
    status: TransferReviewStatusSchema.default('suggested'),
    owner_id: z.number().int().positive().optional(),
    account_id: z.number().int().positive().optional(),
    q: z.string().optional(),
    page: z.number().int().positive().default(1),
    page_size: z.number().int().positive().max(100).default(50)
  })
  .strict()

export type TransferReviewFilters = z.infer<typeof TransferReviewFiltersSchema>

const TransferPairSchema = z
  .object({
    outgoing: TransactionSchema,
    outgoing_owner: UserSummarySchema,
    incoming: TransactionSchema,
    incoming_owner: UserSummarySchema,
    day_gap: z.number().int().nonnegative(),
    match_status: z.enum(['unique', 'ambiguous'])
  })
  .strict()

export const TransferReviewPageSchema = z
  .object({
    items: z.array(TransferPairSchema),
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total: z.number().int().nonnegative(),
    total_pages: z.number().int().positive()
  })
  .strict()

export type TransferReviewPage = z.infer<typeof TransferReviewPageSchema>
export type TransferPair = z.infer<typeof TransferPairSchema>

const TransferReviewUpdateSchema = z
  .object({
    outgoing_transaction_id: z.number().int().positive(),
    incoming_transaction_id: z.number().int().positive(),
    action: z.enum(['confirm', 'reject', 'reset'])
  })
  .strict()

export const TransferReviewUpdateRequestSchema = z
  .object({ items: z.array(TransferReviewUpdateSchema).min(1).max(500) })
  .strict()

export type TransferReviewUpdateRequest = z.infer<typeof TransferReviewUpdateRequestSchema>

export const TransferReviewUpdateResponseSchema = z
  .object({ updated: z.number().int().nonnegative() })
  .strict()

File: backend/app/api/review.py
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.transactions import transaction_response
from app.auth.dependencies import get_current_user, require_csrf
from app.db import request_session
from app.db.models import User
from app.schemas.accounts import UserSummary
from app.schemas.matching import rule_match_response
from app.schemas.transactions import (
    AssignmentBulkUpdate,
    AssignmentBulkUpdateResponse,
    AssignmentReviewPageResponse,
    AssignmentReviewQuery,
    AssignmentReviewRowResponse,
    PatternPreviewExampleResponse,
    PatternPreviewRequest,
    PatternPreviewResponse,
)
from app.schemas.transfers import (
    TransferPairResponse,
    TransferReviewPageResponse,
    TransferReviewQuery,
    TransferReviewUpdateRequest,
    TransferReviewUpdateResponse,
)
from app.services.reviews import (
    bulk_update_assignments,
    get_assignment_review_page,
    preview_patterns,
)
from app.services.transfers import (
    TransferReviewChange,
    get_transfer_review_page,
    review_transfer_pairs,
)

router = APIRouter(prefix="/transactions/review", tags=["assignment-review"])
CurrentUser = Annotated[User, Depends(get_current_user)]
CsrfUser = Annotated[User, Depends(require_csrf)]
DatabaseSession = Annotated[Session, Depends(request_session)]


def _user_summary(user: User) -> UserSummary:
    return UserSummary(
        id=user.id,
        username=user.username,
        first_name=user.first_name,
        last_name=user.last_name,
        is_superuser=user.is_superuser,
    )


@router.get("/transfers", response_model=TransferReviewPageResponse)
def transfer_review_list(
    current_user: CurrentUser,
    session: DatabaseSession,
    query: Annotated[TransferReviewQuery, Query()],
) -> TransferReviewPageResponse:
    result = get_transfer_review_page(
        session,
        current_user,
        status=query.status,
        owner_id=query.owner_id,
        account_id=query.account_id,
        search_term=query.q,
        page=query.page,
        page_size=query.page_size,
    )
    return TransferReviewPageResponse(
        items=[
            TransferPairResponse(
                outgoing=transaction_response(item.outgoing),
                outgoing_owner=_user_summary(item.outgoing.bank_account.owner),
                incoming=transaction_response(item.incoming),
                incoming_owner=_user_summary(item.incoming.bank_account.owner),
                day_gap=item.day_gap,
                match_status=item.match_status,
            )
            for item in result.items
            if item.outgoing.bank_account is not None and item.incoming.bank_account is not None
        ],
        page=result.page,
        page_size=result.page_size,
        total=result.total,
        total_pages=result.total_pages,
    )


@router.patch("/transfers", response_model=TransferReviewUpdateResponse)
def transfer_review_update(
    payload: TransferReviewUpdateRequest,
    current_user: CsrfUser,
    session: DatabaseSession,
) -> TransferReviewUpdateResponse:
    updated = review_transfer_pairs(
        session,
        current_user,
        tuple(
            TransferReviewChange(
                outgoing_transaction_id=item.outgoing_transaction_id,
                incoming_transaction_id=item.incoming_transaction_id,
                action=item.action,
            )
            for item in payload.items
        ),
    )
    return TransferReviewUpdateResponse(updated=updated)


@router.get("", response_model=AssignmentReviewPageResponse)
def assignment_review_list(
    current_user: CurrentUser,
    session: DatabaseSession,
    query: Annotated[AssignmentReviewQuery, Query()],
) -> AssignmentReviewPageResponse:
    result = get_assignment_review_page(
        session,
        current_user,
        issue=query.issue,
        owner_id=query.owner_id,
        account_id=query.account_id,
        contract_id=query.contract_id,
        search_term=query.q,
        page=query.page,
        page_size=query.page_size,
    )
    return AssignmentReviewPageResponse(
        items=[
            AssignmentReviewRowResponse(
                transaction=transaction_response(item.transaction),
                account_name=item.account.name,
                owner=UserSummary(
                    id=item.account.owner.id,
                    username=item.account.owner.username,
                    first_name=item.account.owner.first_name,
                    last_name=item.account.owner.last_name,
                    is_superuser=item.account.owner.is_superuser,
                ),
                category_match=rule_match_response(item.category_match),
                contract_match=rule_match_response(item.contract_match),
                issues=list(item.issues),
            )
            for item in result.items
        ],
        page=result.page,
        page_size=result.page_size,
        total=result.total,
        total_pages=result.total_pages,
    )


@router.patch("", response_model=AssignmentBulkUpdateResponse)
def assignment_review_update(
    payload: AssignmentBulkUpdate,
    current_user: CsrfUser,
    session: DatabaseSession,
) -> AssignmentBulkUpdateResponse:
    updated = bulk_update_assignments(
        session,
        current_user,
        tuple(dict.fromkeys(payload.transaction_ids)),
        set_category=payload.set_category,
        category_id=payload.category_id,
        set_contract=payload.set_contract,
        contract_id=payload.contract_id,
    )
    return AssignmentBulkUpdateResponse(updated=len(updated))


@router.post("/patterns/preview", response_model=PatternPreviewResponse)
def assignment_pattern_preview(
    payload: PatternPreviewRequest,
    current_user: CsrfUser,
    session: DatabaseSession,
) -> PatternPreviewResponse:
    result = preview_patterns(
        session,
        current_user,
        payload.patterns,
        owner_id=payload.owner_id,
        start_date=payload.start_date,
        end_date=payload.end_date,
    )
    return PatternPreviewResponse(
        total=result.total,
        examples=[
            PatternPreviewExampleResponse(
                id=transaction.id,
                recipient=transaction.recipient,
                subject=transaction.subject,
                date_issue=transaction.date_issue,
                account_name=account.name,
            )
            for transaction, account in result.examples
        ],
    )
